### PR TITLE
avoid PermissionError on Lustre when copying files (fixes #25354)

### LINKED ISSUE #25354: Installation issue: openjdk
### Steps to reproduce the issue

```
$ spack install openjdk@11.0.8_10
==> Warning: gcc@8.4.1 cannot build optimized binaries for "cascadelake". Using best target possible: "skylake_avx512"
==> Installing openjdk-11.0.8_10-ezp6p5txswxryvypbhmw7zsqm45yn2kd
==> No binary for openjdk-11.0.8_10-ezp6p5txswxryvypbhmw7zsqm45yn2kd found: installing from source
==> Using cached archive: /g/data/jh2/bw0729/spack/var/spack/cache/_source-cache/archive/6e/6e4cead158037cb7747ca47416474d4f408c9126be5b96f9befd532e0a762b47.tar.gz
==> No patches needed for openjdk
==> openjdk: Executing phase: 'install'
==> Error: PermissionError: [Errno 13] Permission denied: '/g/data/jh2/bw0729/spack/opt/spack/linux-centos8-skylake_avx512/gcc-8.4.1/openjdk-11.0.8_10-ezp6p5txswxryvypbhmw7zsqm45yn2kd/legal/java.smartcardio/pcsclite.md'

/g/data/jh2/bw0729/spack/var/spack/repos/builtin/packages/openjdk/package.py:123, in install:
        121    def install(self, spec, prefix):
        122        top_dir = 'Contents/Home/' if platform.system() == "Darwin" else '.'
  >>    123        install_tree(top_dir, prefix)

See build log for details:
  /scratch/jh2/bw0729/tmp/spack-stage/spack-stage-openjdk-11.0.8_10-ezp6p5txswxryvypbhmw7zsqm45yn2kd/spack-build-out.txt
```

### Information on your system

* **Spack:** 0.16.2-3870-5916afec84
* **Python:** 3.6.8
* **Platform:** linux-centos8-cascadelake
* **Concretizer:** original

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/6965940/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/6965941/spack-build-env.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by BenWibking**
There is a bug in Python versions < 3.7.4 that causes PermissionError to occur when using shutil.copy2 to copy read-only files that have extended attributes set by Lustre (which is used by lib/spack/llnl/util/filesystem.py:504 to copy directory trees). The upstream Python bug is [here](https://bugs.python.org/issue24538).

Setting SPACK_PYTHON to point to a python executable >= version 3.7.4 fixes the problem. However, for Python versions below this, we monkeypatch shutil.copystat to set extended attributes *before* setting permissions. This means that copying read-only files that have extended attributes set by Lustre can succeed, which is necessary to install openjdk (and fix #25354). I have confirmed on my cluster that this patch fixes openjdk installation when spack is installed on a Lustre FS when using Python 3.6.

### CODE REVIEW

**Comment by BenWibking** [comment]
It looks like the CI fails due to this function accessing a member variable that only exists in Python 3.3+, but I've guarded against accessing it for earlier verions:
 ```
           # use the real function only if it exists
            # *and* it supports follow_symlinks
            def lookup(name):
                fn = getattr(os, name, _nop)
                if sys.version_info >= (3, 3):
                    if fn in os.supports_follow_symlinks:
                        return fn
                return _nop
```

**Comment by mwkrentel** [comment]
I tried this on theta at ANL.  `/lus/theta-fs0` is lustre.

```
$ mount | grep  lustre
172.22.10.92@o2ib:172.22.10.93@o2ib:/snx11214 on /lus/theta-fs0 type lustre (rw,flock,lazystatfs)
```

On `/lus/theta-fs0`, current spack develop fails for me as:

```
$ spack install openjdk 
==> Installing openjdk-11.0.12_7-cw7hhkrfe4x2fk73jlxw22pefvf5n4j5
==> No binary for openjdk-11.0.12_7-cw7hhkrfe4x2fk73jlxw22pefvf5n4j5 found: installing from source
==> Using cached archive: /lus/theta-fs0/projects/CSC250STTO11/krentel/spack/spack-repo/source-cache/_source-cache/archive/87/8770f600fc3b89bf331213c7aa21f8eedd9ca5d96036d1cd48cb2748a3dbefd2.tar.gz
==> No patches needed for openjdk
==> openjdk: Executing phase: 'install'
==> Error: PermissionError: [Errno 13] Permission denied: '/lus/theta-fs0/projects/CSC250STTO11/krentel/spack/install/cray-cnl7-x86_64/gcc-11.2.0/openjdk-11.0.12_7-cw7hhkrfe4x2fk73jlxw22pefvf5n4j5/legal/jdk.localedata/thaidict.md'

/lus/theta-fs0/projects/CSC250STTO11/krentel/spack/spack-repo/var/spack/repos/builtin/packages/openjdk/package.py:144, in install:
        142    def install(self, spec, prefix):
        143        top_dir = 'Contents/Home/' if platform.system() == "Darwin" else '.'
  >>    144        install_tree(top_dir, prefix)
```

With the patch to `filesystem.py` it works.

```
$ spack install openjdk
==> Installing openjdk-11.0.12_7-cw7hhkrfe4x2fk73jlxw22pefvf5n4j5
==> No binary for openjdk-11.0.12_7-cw7hhkrfe4x2fk73jlxw22pefvf5n4j5 found: installing from source
==> Using cached archive: /lus/theta-fs0/projects/CSC250STTO11/krentel/spack/spack-repo/source-cache/_source-cache/archive/87/8770f600fc3b89bf331213c7aa21f8eedd9ca5d96036d1cd48cb2748a3dbefd2.tar.gz
==> No patches needed for openjdk
==> openjdk: Executing phase: 'install'
==> openjdk: Successfully installed openjdk-11.0.12_7-cw7hhkrfe4x2fk73jlxw22pefvf5n4j5
  Fetch: 0.49s.  Build: 6.69s.  Total: 7.18s.
[+] /lus/theta-fs0/projects/CSC250STTO11/krentel/spack/install/cray-cnl7-x86_64/gcc-11.2.0/openjdk-11.0.12_7-cw7hhkrfe4x2fk73jlxw22pefvf5n4j5
```

So, I can confirm the bug on lustre and the patch works for me.

Btw, it's not just the `install_tree` and `module_roots` directories
where it fails, it also fails when `build_stage` is on lustre.

This is with current spack develop as of today, Nov 9, 2021.

* **Spack:** 0.17.0-48-0322431c5c
* **Python:** 3.6.12
* **Platform:** cray-sles15-haswell
* **Concretizer:** clingo

**Comment by mwkrentel** [comment]
I tested this a week ago, and I was able to reproduce the problem and
confirm the solution works.  So, I approve.

@tldahlgren You understand I don't have commit access (and don't really want it), right?

**Comment by tldahlgren** [comment]
> I tested this a week ago, and I was able to reproduce the problem and confirm the solution works. So, I approve.
> 
> @tldahlgren You understand I don't have commit access (and don't really want it), right?

Yes, I do understand.  

I was just flagging PRs with the people who have provided reviews so it's visible on the PR list.  IMO it's an easier way to distinguish PRs with real feedback (e.g., something other than spackbot comments).

Perhaps this is a good topic to clarify in the Spack User meeting?

**Comment by BenWibking** [comment]
@tgamblin @becker33 Is there any timeframe on when this can get reviewed? This patch fixes an issue I reported over 3 months ago, and this PR itself is 20 days old...

**Comment by haampie** [comment]
Not sure why this was not reviewed, but one issue is we would add PSF-licensed
code, if I understand correctly.

```
2. Subject to the terms and conditions of this License Agreement, PSF hereby
   grants Licensee a nonexclusive, royalty-free, world-wide license to reproduce,
   analyze, test, perform and/or display publicly, prepare derivative works,
   distribute, and otherwise use Python 3.10.0 alone or in any derivative
   version, provided, however, that PSF's License Agreement and PSF's notice of
   copyright, i.e., "Copyright © 2001-2021 Python Software Foundation; All Rights
   Reserved" are retained in Python 3.10.0 alone or in any derivative version
   prepared by Licensee.

3. In the event Licensee prepares a derivative work that is based on or
   incorporates Python 3.10.0 or any part thereof, and wants to make the
   derivative work available to others as provided herein, then Licensee hereby
   agrees to include in any such work a brief summary of the changes made to Python
   3.10.0.
```

Can you rebase the PR?

**Comment by BenWibking** [comment]
Rebased.

I didn't realize the license issue. How does this work for vendored dependencies? Would adding shutil (from 3.7.4 or higher) as a vendored dependency fix this problem?

File: lib/spack/llnl/util/filesystem.py
Comment (haampie) [inline lib/spack/llnl/util/filesystem.py]:
```suggestion
                    if fn in os.supports_follow_symlinks:  # novermin
```

**Comment by BenWibking** [comment]
The error message for the unittests (2.7, clingo) check is inscrutable to me. Is this a CI problem, or an actual problem with this patch?

**Comment by BenWibking** [comment]
@haampie @tgamblin @becker33 Just trying to get traction on getting this merged. Is there anything blocking merging this?

**Comment by BenWibking** [comment]
@haampie @tgamblin @becker33 Checking in on this...any possibility of getting this merged?

**Comment by haampie** [comment]
There's still some issue with Python 2.7, see the failing test.

```
TypeError: stat() takes no keyword arguments
```

**Comment by wyphan** [comment]
@spackbot fix style

**Comment by BenWibking** [comment]
> There's still some issue with Python 2.7, see the failing test.
> 
> ```
> TypeError: stat() takes no keyword arguments
> ```

Since Python 2.7 support has been dropped, is there anything currently blocking this?

**Comment by wyphan** [comment]
I tried rebasing to latest develop, and it succeeded. But it turns out I
don't have write access to your branch 🥲

**Comment by BenWibking** [comment]
I've checked the "Allow edits and access to secrets by maintainers" box. Is there anything else I need to do?

**Review by haampie — APPROVED** [review]

**Pull request merged**
